Extended Clustermesh (Clustermesh511)

thorn3r opened the pull request:
Support extending Clustermesh to > 255 clusters

This PR introduces a new flag `--max-connected-clusters` which allows the bit allocation of Identity/ClusterID in a NumericIdentity to be configured to allow connecting up to 511 clusters (9 bits) in a Clustermesh, at the expense of shrinking the allocatable cluster-local identities to 32767 (15 bits).

The ClusterID field of 5 BPF maps is extended to uint16 to accomodate larger ClusterIDs:
- cilium_lxc
- cilium_lb4_backends_v3
- cilium_lb6_backends_v3
- cilium_tunnel_map
- cilium_ipcache

Map migrations were not necessary for anys maps due to:
- Only consuming padding (lxc and lb_backends)
- The maps being recreated on agent start (tunnel and ipcache maps)

Currently the only supported values for `max-connected-clusters` are 255 (default) and 511.

This feature has no live-migration path and can only (currently) be used for brand new clusters (i.e. you cannot change a 255 cluster to 511).

Additional details available in individual commit messages

Release note:
```release-note
Add support for extending ClusterMesh to 511 clusters
By setting the flag `--max-connected-clusters=511`, a new cluster will be able to connect to a ClusterMesh with up to 511 clusters. If enabled, the number of possible cluster-local identities will be reduced to 32,768. This feature can only be enabled on new clusters, and all clusters in the ClusterMesh must share the same configuration.
```

thorn3r committed:
bpf: Support extended clustermesh option

Extended clustermesh changes the bit allocation in skb->mark to repurpose
bits from identity and grant them to cluster ID. Users can configure
this with the agent flag '--max-connected-clusters'.

This enables support for larger clustermesh/greater cluster ID but
reduces the maximum allocatable identity.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
helm: update charts to include maxConnectedClusters option

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
workflow/clustermesh: set maxConnectedClusters

Set maxConnectedClusters in ci-multicluster to test extended clustermesh feature.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
test: set Cluster ID in upgrade-downgrade test

Setting Cluster ID in the upgrade-downgrade test will provide test
coverage for the updates made to BPF maps and the datapath to provide
support for extended clustermesh. These changes include expanding the
Cluster ID field on lxc, endpoint, lb4backend, and lb6backend maps, as
well as modifications to skb->mark parsing for identity.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
bpf maps: extend cluster_id to uint16

Extend the cluster_id field of the keys/values for the following maps
from uint8 to uint16 to support larger ClusterMesh:
- lxc
- tunnel
- ipcache
- lb_backends

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
operator: enforce MaxConnectedClusters for CiliumClusterConfig

With the addition of MaxConnectedClusters to CiliumClusterConfig, the
operator's KVStore watchdog must be aware of and set this value to ensure
it is not removed on update after updating the heartbeat.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
daemon, clustermesh: Add flag for enabling extended clustermesh

This commit adds a new flag '--max-connected-clusters' to enable
users to configure the bit allocation of cluster ID and identity
in a numeric identity to enable support for a clustermesh with
more than 255 clusters.

Currently the only supported values for this flag are 255 (default)
and 511. Increasing this value will increase the maximum number of
clusters/maximum cluster ID by reducing the maximum number of
allocatable cluster-local identities.

This feature changes the way identites are interpreted in the datapath.
It should only be used for new clusters, as there is currently
no migration path for live clusters. All clusters connected in a
clustermesh will require this flag to be set to the same value.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
clustermesh-apiserver: Add flag for enabling extended clustermesh

This commit adds a new flag '--max-connected-clusters' to
clustermesh-apiserver. The value will be written to the
CiliumClusterConfig and is used by remote-clusters to validate that both
clusters are running in the same identity/clusterID allocation mode
prior to synchronizing resources.

Clusters that are using different values for '--max-connected-clusters'
should never be permitted to connect to each other, as this determines
how identity is parsed by datapath BPF programs.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

giorio94 commented on kvstoremesh/main.go:
AFAIU, the changes introduced in this commit to kvstoremesh are not necessary, and could be dropped (exept for propagating the clusterconfig value). KVStoreMesh only propagates verbatim the information from remote clusters to the local etcd, and it does not need to be involved with checking the configuration of the number of clusters/identities.

thorn3r replied:
ah yes, you're right. I misunderstood the context of where/when clustermesh/remoteCluster.Run vs kvstoremesh/remoteCluster.Run are executed.

Since clustermesh/remoteCluster.Run is executed in the daemon for both kvstoremesh and non-kvstoremesh, we can drop all but the cluster config propagation and move the maxConnectedClusters validation into the config validation as you suggested, I like this :+1:

thorn3r replied:
So turns out we _do_ actually need some of these changes. Without kvstoremesh properly validating the ClusterID and MaxConnectedClusters, it's possible a user could add an invalid cluster to a kvstoremesh and the operator would still write the config to the kvstore and begin synchronization.

This means the validation should probably happen in `restartRemoteConnection()` before we execute `rc.Run()`, so that both controllers in the agent and kvstoremesh-operator will perform the validation. Unless we are ok with the kvstoremesh synchronization if the agents themselves are not watching the prefixes for the remote cluster?

giorio94 replied:
Yeah, I personally think that it would be OK to have kvstoremesh synchronize these entries and then the agents failing to watch them if there's a mismatch. Currently, the same is also happening for the ClusterID validation, which is performed by the agents only, so that kvstoremesh doesn't have to deal with which features are enabled (and theoretically we could use kvstoremesh for some remote clusters but not others, so it is also impossible to perform uniqueness checks for instance).

This way, we have all failures reported in the same place (the agents) for both standard clustermesh-apiserver and kvstoremesh. At the end of the day, that can occur only if there's a configuration mismatch that will be eventually need resolved by an operator, so I would not be concerned by the possible performance overhead due to the unnecessary synchronization.

thorn3r replied:
Yeah I like that reasoning, makes sense to me :+1:

giorio94 commented on pkg/clustermesh/common/remote_cluster.go:
```diff
@@ -189,6 +192,11 @@ func (rc *remoteCluster) restartRemoteConnection() {
+					rc.getLogger().WithError(err).Error("Terminating remote cluster connection due to invalid remote cluster configuration")
+					rc.releaseOldConnection()
+					return err
+				}
```
I'd personally include this validation along that of the ClusterID for consistency (https://github.com/cilium/cilium/blob/eb1f266a5da54d19300e38a7e9a48c2d53852b89/pkg/clustermesh/remote_cluster.go#L68-L72), and perform it only in the clustermesh case (as kvstoremesh doesn't really need it).

giorio94 commented on pkg/clustermesh/common/remote_cluster.go:
```diff
@@ -57,6 +57,9 @@ type remoteCluster struct {
 
+	// mesh is the cluster mesh this remote cluster belongs to
+	mesh *ClusterMesh
+
```
I'd personally avoid introducing the circular dependency on ClusterMesh here, unless strictly necessary. I know that we already have it in the `pkg/clustermesh/remote_cluster` structure, but it looks like a code smell and it makes more difficult to figure out which fields are actually accessed.

thorn3r replied:
I didn't feel great about this either, but we won't need this after I move the validation to clustermesh/remoteCluster.Run :+1:

giorio94 replied:
Yeah, this should not be necessary anymore.

giorio94 replied:
This could be reverted, as no longer used.

giorio94 reviewed (commented):
This looks really cool! I quickly skimmed through the clustermesh changes, and left a couple of minor comments inline about possible simplifications.

kaworu reviewed (approved):
Helm changes LGTM.

thorn3r committed:
kvstoremesh-operator: write MaxConnectedClusters to config

The field MaxConnectedClusters was added to CiliumClusterConfig in a
previous commit to support the extended Clustermesh feature. This value
must be written to cluster configs in the kvstore to ensure that all
clusters in a Clustermesh are using the same identity/clusterID
allocation more prior to synchronization.

Clusters that are using different values for '--max-connected-clusters'
should never be permitted to connect to each other, as this determines
how identity is parsed by datapath BPF programs.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

aanm commented on pkg/identity/numericidentity.go:
```diff
@@ -66,6 +63,13 @@ const (
+
+	// ClusterIDShift is the number of bits to shift a cluster ID in a numeric
+	// identity and is equal to the number of bits that represent a cluster-local identity.
+	ClusterIDShift int = NumericIdentityBitlength - ClusterIDLen
```
I would suggest to have Getters and Setters for these variables and then [1]

aanm commented on pkg/identity/numericidentity.go:
```diff
@@ -365,6 +369,15 @@ func InitWellKnownIdentities(c Configuration) int {
+	// ClusterIDLen is the number of bits that represent a cluster ID in a numeric identity
+	ClusterIDLen = int(math.Log2(float64(cmtypes.ClusterIDMax + 1)))
+	// ClusterIDShift is the number of bits to shift a cluster ID in a numeric identity
+	ClusterIDShift = NumericIdentityBitlength - ClusterIDLen
```
[1] in here we should panic in case the function `InitClusterIDShift` is ever called after we have done a "get" on either "ClusterIDLen" or "ClusterIDShift".

thorn3r replied:
rather than a panic, i added Getters that initialize the variables the 1st time one of them is used. what do you think about this approach? (just realized i need to update the commit msg)

aanm commented on bpf/lib/overloadable_skb.h:
```diff
@@ -43,8 +48,12 @@ get_epid(const struct __sk_buff *ctx)
+	__u32 cluster_id_upper = ((cluster_id & 0xFFFFFF00) << (8 + IDENTITY_LEN));
+
+	ctx->mark &= MARK_MAGIC_KEY_MASK;
+	ctx->mark |= (identity & IDENTITY_MAX) << 16 | cluster_id_lower | cluster_id_upper;
```
I think an ascii diagram representing the bits meaning would be useful.

thorn3r replied:
added :+1:

aanm commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -2708,6 +2708,9 @@ clustermesh:
     #     caCert: ""
+    # -- The maximum number of clusters to support in a clustermesh.
+    # Values > 255 will decrease maximum allocatable identities per cluster.
+    maxConnectedClusters: 255
```
Can this value be 256? Can it be 257? Can it be 510? Can it be 517? It's a hint that maybe we should have some helm validation in this field :smile:

giorio94 replied:
Two aspects in addition to André's comment here:

* I'd explicitly state which are the supported values;
* How would this play with the `config.enabled` flag above? More in general, I wonder if is the correct place for this setting, as `config` here currently represents the list of remote clusters. For instance, the cluster ID and cluster name settings, which are also clustermesh-related, are in the global agent section. Should this also be there? Should we even not expose this through helm, and make it configurable only through the `extraConfig` section? This option might make sense if we expect this to be used only in very niche subset of scenarios, and if properly documented.

thorn3r replied:
good point on the config aspect. I moved `maxConnectedClusters` to `clustermesh.maxConnectedClusters` and added helm validation.

@giorio94 I think your suggestion about extraConfig also makes sense, given that this might not be a widely used feature. If either of you think this is a better fit, I can look at making that change

aanm reviewed (changes requested):
I did an overall review. I relied on Marco to look much closer at the clustermesh bits.

giorio94 commented on api/v1/openapi.yaml:
```diff
@@ -2583,6 +2583,9 @@ definitions:
         type: boolean
+      max-connected-clusters:
+        description: The maximum number of clusters this cluster can support in a Clustermesh
+        type: integer
```
Do we need this information for every cluster, or should it be something global for the entire clustermesh? I mean, by constraint, this number must be the same for all clusters, otherwise the connection gets reset, so it seems a bit redundant. Additionally, it seems that this filed is currently set, but never printed.

thorn3r replied:
You're right, this was unnecessary. I removed it :+1:

giorio94 commented on daemon/cmd/daemon_main.go:
```diff
@@ -1128,6 +1129,10 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	option.BindEnv(vp, option.EnableK8sNetworkPolicy)
 
+	flags.Uint32(option.MaxConnectedClusters, defaults.MaxConnectedClusters, "Maximum number of clusters to be connected in a clustermesh. Increasing this value will reduce the maximum number of identities available.")
+	flags.MarkHidden(option.MaxConnectedClusters)
```
Is it there a specific reason for marking this flag hidden? Especially given that in a later commit the same setting is explicitly provided through Helm?

giorio94 commented on daemon/cmd/daemon_main.go:
```diff
@@ -1128,6 +1129,10 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	flags.MarkHidden(option.EnableK8sNetworkPolicy)
 	option.BindEnv(vp, option.EnableK8sNetworkPolicy)
 
+	flags.Uint32(option.MaxConnectedClusters, defaults.MaxConnectedClusters, "Maximum number of clusters to be connected in a clustermesh. Increasing this value will reduce the maximum number of identities available.")
```
Nit. It could make sense to explicitly state the allowed values in the flag description.

giorio94 commented on pkg/clustermesh/common/clustermesh.go:
The changes in this file could be reverted, as no longer used.

giorio94 commented on pkg/clustermesh/types/types.go:
```diff
@@ -4,18 +4,33 @@
 )
 
+// ClusterIDMax is the maximum value of the cluster ID
+var ClusterIDMax uint32 = defaults.MaxConnectedClusters
```
Not sure how feasible it is in this case without an extensive refactoring, but we should strive to avoid using run-time initialized global variables as much as possible. They tend to be problematic especially in combination with the hive framework, as depending on the initialization order of the cells they might, or might not be already correctly initialized (I see that you are initializing this before starting the hive, which is safe, but there's still an ongoing effort to move the registration of the flags to the appropriate places, which might change it).

In this specific case, I'd suggest wrapping this in an object which can be used to register the flags (so that it conforms with he Hive approach, and it can also be reused for the clustermesh-apiserver) and perform the validation. This way, the components which are already cells (e.g., clustermesh) can depend on it, while the others can still rely on the global variable to prevent an extensive refactoring. The same object could also be used to do the validation of the `ClusterConfig`, relieving the main ClusterMesh object from that task,

thorn3r replied:
ah yes, I like this idea. I already began a refactor to remove this global, but I had planned on replacing it just with usage of the DaemonConfig field. I like your idea better

thorn3r replied:
in lieu of removing the `ClusterIDMax` global entirely (I began this several times, but the refactor ballooned. Probably best for its own PR), I've moved maxConnectedClusters into `ClusterInfo` , which validates and populates `ClusterIDMax` to be used in places that haven't been modularized yet.

giorio94 commented on pkg/clustermesh/types/types.go:
```diff
@@ -29,7 +44,8 @@ func ValidateClusterID(clusterID uint32) error {
 type CiliumClusterConfig struct {
-	ID uint32 `json:"id,omitempty"`
+	ID                   uint32 `json:"id,omitempty"`
+	MaxConnectedClusters uint32 `json:"maxConnectedClusters,omitempty"`
```
I wonder if this would fit more into the `Capabilities` section below.

thorn3r replied:
I think so too, I moved it there

giorio94 commented on pkg/clustermesh/clustermesh.go:
```diff
@@ -212,3 +217,16 @@ func (cm *ClusterMesh) Status() (status *models.ClusterMeshStatus) {
 	return
 }
+
+func (cm *ClusterMesh) ExtendedClustermeshEnabled() bool {
```
Nit. ClusterMesh is typically spelled with the capital `M`.

giorio94 commented on pkg/clustermesh/kvstoremesh/remote_cluster.go:
```diff
@@ -56,6 +57,7 @@ func (rc *remoteCluster) Run(ctx context.Context, backend kvstore.BackendOperati
 	var capabilities types.CiliumClusterConfigCapabilities
 	if srccfg != nil {
 		capabilities = srccfg.Capabilities
+		dstcfg.MaxConnectedClusters = srccfg.MaxConnectedClusters
```
Nit. This change is redundant.

giorio94 commented on Documentation/spelling_wordlist.txt:
```diff
@@ -451,6 +452,7 @@ mainboard
 makefile
 masq
 matchPattern
+maxConnectedClusters
```
This should not be necessary, as helm keys are automatically skipped from spell checking.

zacharysarah replied:
@thorn3r Are you seeing build failures on this term?

thorn3r replied:
@zacharysarah I was early on, but it's possible that was before I finished the helm changes. I'll reassess and remove this if its not needed

thorn3r replied:
removed this from the wordlist!

giorio94 commented on test/k8s/updates.go:
```diff
@@ -17,6 +17,11 @@ import (
 
+const (
+	clusterName = "test-upgrade-downgrade"
+	clusterID   = "1"
```
Nit. Maybe a larger value (for instance `170`, which translates into `10101010` in binary) would provide an even better coverage.

giorio94 reviewed (changes requested):
I took another pass on the clustermesh/agent modifications, and overall the approach looks good to me. Mostly nits inline (marking as requested changes only because plain comment doesn't play well with the review queue).

ldelossa commented on bpf/lib/common.h:
```diff
@@ -304,8 +304,8 @@ struct tunnel_key {
 	__u8 family;
-	__u8 cluster_id;
-	__u16 pad;
+	__u8 pad;
```
Is there a particular reason we put pad in the middle, and don't do

```c
struct tunnel_key {
	union {
		struct {
			__u32		ip4;
			__u32		pad1;
			__u32		pad2;
			__u32		pad3;
		};
		union v6addr	ip6;
	};
	__u8 family;
        __u16 cluster_id;
	__u8 pad;
} __packed;
```

The structure is packed so we don't have to worry about any implicit padding I think.

thorn3r replied:
For the C struct this would be fine, but the mirrored Go struct isn't packed and throws off the alignment:
```
--- FAIL: TestCheckStructAlignments (0.00s)
    alignchecker_test.go:16: type TunnelKey's binary.Size (20) does not equal its unsafe.Sizeof (22) size (struct with implicit trailing padding?)
FAIL
```

ldelossa commented on bpf/lib/overloadable_skb.h:
```diff
@@ -25,7 +26,11 @@ bpf_clear_meta(struct __sk_buff *ctx)
 static __always_inline __maybe_unused int
 get_identity(const struct __sk_buff *ctx)
```
Since these are changing now, it maybe cool to try to implement a BPF unit tests for the functions you change, assuming there aren't limitations to doing so.

thorn3r replied:
Good suggestion! I added some unit tests in the most recent commit

ldelossa reviewed (approved):
changes LGTM.

ti-mo commented on bpf/cluster_config.h:
For https://github.com/cilium/cilium/issues/27446, we'll eventually need to remove all runtime-generated `*_config.h` files. Could the values here be kept in `node_config.h`? (it's already a node-wide header after all, included by every entrypoint) That way, we'll have one less header to phase out and you'll have less boilerplate to worry about.

thorn3r replied:
that was my initial approach here, but ran into too many issues trying to include node_config.h in common.h.

I think alternatively we could roll these into node_config.h, and add a new `clustermesh_common.h` or similar and include that where needed. Which approach do you prefer?

thorn3r replied:
Since cluster_config.h should be immutable now (see https://github.com/cilium/cilium/pull/27520/commits/bdc4c8a20e166bcf7360a18e908a6541f0e18c1f), it seems like it's better to keep this separate from node_config.h

ti-mo replied:
I'm a bit out of the loop with the developments that happened in the past month. Wouldn't it be simpler to remove the writable bit from cluster_config.h if we want to prevent future modifications? If we want certain configuration values to never change over the lifetime of a cluster, I'd suggest we find a solution on a higher level instead. (Do we have validating webhooks for configmap updates?)

Your proposal effectively validates user configuration by reading a header file. I think that's going to prove fragile and it paints us into a corner. As I mentioned before, I want to start moving `#define`s to runtime-injected constants, so removing `cluster_config.h` would also mean removing the validation mechanism, since there will be no more state on disk.

Could you outline the exact issues you ran into adding these values to node_config.h? If necessary, you could include them closer to where they're used instead of blanket common.h.

zacharysarah reviewed (changes requested):
@thorn3r 👋🏻 If you're seeing build failures for `maxConnectedClusters`, then it's OK to add it--otherwise, let's [remove it per](https://github.com/cilium/cilium/pull/27520/files?file-filters%5B%5D=.go&file-filters%5B%5D=.h&file-filters%5B%5D=.md&file-filters%5B%5D=.rst&file-filters%5B%5D=.tmpl&file-filters%5B%5D=.txt&file-filters%5B%5D=.yaml&owned-by%5B%5D=zacharysarah&show-deleted-files=true&show-viewed-files=true#r1307024979) @giorio94. Please re-request review when you're ready.

thorn3r committed:
bpf: add unit tests for skb funcs

New unit tests have been added to cover changes made to the skb
functions that get and set identity and cluster_id in skb->mark.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

zacharysarah reviewed (approved):
LGTM

giorio94 commented on pkg/clustermesh/types/option.go:
```diff
@@ -17,27 +17,34 @@ const (
 func (def ClusterInfo) Flags(flags *pflag.FlagSet) {
 	flags.Uint32(OptClusterID, def.ID, "Unique identifier of the cluster")
 	flags.String(OptClusterName, def.Name, "Name of the cluster")
+	flags.Uint32(OptMaxConnectedClusters, def.MaxConnectedClusters, "Maximum number of clusters to be connected in a clustermesh. Increasing this value will reduce the maximum number of identities available")
```
I'd explicitly state the allowed values in the description.

giorio94 commented on pkg/clustermesh/types/option.go:
```diff
@@ -17,27 +17,34 @@ const (
 
-// Validate validates that the ClusterID is in the valid range (including ClusterID == 0),
+// Validate validates that MaxConnectedClusters is configured with a supported
+// value, that the ClusterID is in the valid range (including ClusterID == 0),
```
The comment change is not reflected in the code.

giorio94 commented on pkg/clustermesh/types/types.go:
```diff
@@ -4,18 +4,33 @@
+// ClusterIDMax is the maximum value of the cluster ID
+var ClusterIDMax uint32 = defaults.MaxConnectedClusters
+
+// InitClusterIDMax validates and sets the ClusterIDMax package level variable.
+func (c ClusterInfo) InitClusterIDMax() error {
+	switch c.MaxConnectedClusters {
+	case defaults.MaxConnectedClusters, ClusterIDExt511:
+		ClusterIDMax = c.MaxConnectedClusters
+	default:
+		return fmt.Errorf("--%s=%d is invalid; supported values are [%d, %d]", OptMaxConnectedClusters, c.MaxConnectedClusters, defaults.MaxConnectedClusters, ClusterIDExt511)
+	}
+	return nil
+}
```
Any reason not execute this as part of `ClusterInfo.Validate` (or `InitAndValidate`)?

thorn3r replied:
I initially had it in `Validate`, but it didn't feel right initializing variables in a validation method. I can combine into a new `InitAndValidate` if you prefer that approach

giorio94 replied:
My though was just that at a certain point the variable initialization would be (hopefully) dropped, and we'll keep only the validation side, but I see your point. I don't have strong opinions honestly here.

giorio94 commented on pkg/option/config.go:
```diff
@@ -1204,6 +1204,13 @@ const (
+	// The value is used to determine the bit allocation for cluster ID and
+	// identity in a numeric identity. Values > 255 will decrease the number of
+	// allocatable identities.
+	MaxConnectedClusters = "max-connected-clusters"
```
Should be reverted, as no longer used.

giorio94 commented on pkg/clustermesh/clustermesh.go:
```diff
@@ -240,3 +242,16 @@ func (cm *ClusterMesh) Status() (status *models.ClusterMeshStatus) {
+	}
+
+	return nil
+}
```
Does this need to be a public ClusterMesh method? It seems more related to the ClusterInfo struct (like `IsCompatible(...)`, or something like that). It could also include the entire validation logic, which is now part of RemoteCluster.

thorn3r replied:
I think this (moving all into ClusterInfo.ValidateRemote or something similar) will require moving `ConfigValidationMode` to the ClusterInfo struct as well, otherwise it will be more difficult to test this aspect. Does this seems alright?

giorio94 replied:
Hmm, couldn't that be a parameter?

thorn3r replied:
updated with your feedback

giorio94 commented on pkg/maps/nat/per_cluster_nat.go:
```diff
@@ -20,9 +20,10 @@ const (
-	perClusterMapMaxEntries  = cmtypes.ClusterIDMax + 1
 )
 
+var perClusterMapMaxEntries = cmtypes.ClusterIDMax + 1
```
This change will not work, as the updated ClusterIDMax value would not be reflected here: https://go.dev/play/p/-O8-4pZbagj. I'd drop this global variable, and pass the `ClusterIDMax` as parameter to `NewPerClusterCTMaps`, adapting the tests accordingly. Same for the ctmap occurrence.

thorn3r replied:
wow good catch, didn't even notice this. thank you

giorio94 commented on bpf/lib/common.h:
```diff
@@ -722,7 +722,9 @@ enum metric_dir {
 
 /* Shouldn't interfere with MARK_MAGIC_TO_PROXY. Lower 8bits carries cluster_id */
```
This comment does no longer seem accurate.

giorio94 commented on bpf/cluster_config.h:
```diff
@@ -0,0 +1,22 @@
+#define CLUSTER_ID_LEN 8
+#define CLUSTER_ID_MAX 255
+#define IDENTITY_LEN 16
+#define IDENTITY_MAX 65535
```
X_LEN and X_MAX are directly related to each other, and given one it is possible to compute the other. Is it necessary to actually write both of them in the header file, or could we just use an helper to compute them? Something like `#define MAX_VALUE(LEN) ((1<<LEN)-1)`.

thorn3r replied:
we don't need to write them to the config, but I thought its usage is a little more concise (using `CLUSTER_ID_MAX` rather than `CLUSTER_ID_MAX(CLUSTER_ID_LEN)`.
What do you think about defining the MAX macros in common.h instead, and only writing the LEN defines to the header?

thorn3r replied:
actually, it may not make sense to do this in the way i described. the maximum cluster ID is what is actually passed in via flag. The len of each is calculated once using the Getters added in [this commit](https://github.com/cilium/cilium/pull/27520/commits/ac6778cb54812db8ce35c84fae03bf541f88f18e). I think it's not explicitly necessary to define all 4 in the header, but it seems a bit simpler

giorio94 replied:
I don't have a strong opinion, it just seemed a bit redundant to specify both `X_LEN` and `X_MAX` as part of the config. But if that is more convenient, then it is fine for me.

giorio94 commented on pkg/datapath/linux/config/config.go:
```diff
@@ -140,6 +132,48 @@ func (h *HeaderfileWriter) WriteClusterConfig(w io.Writer) error {
+	}
+
+	return nil
+}
```
What would happen to the config files in case Cilium is uninstalled? I mean, I guess that they would be left on disk, but this check would prevent changing the setting even with a full re-installation of Cilium.

thorn3r replied:
I think this will only be a problem if the user uninstalled Cilium and tried to re-install with a different setting for `max-connected-clusters`. Should I add a message to the error like "If this is a brand-new installation, remove the cluster_config.h file and try again`? The surface area seems pretty small, especially since this validation can probably be removed one the migration is implemented, but I'm not sure how else to handle this

giorio94 replied:
Yeah, I agree that it would only occur for re-installations changing that parameter, but that would still be a possibility (especially because the block is to prevent changing it while cilium is installed, so there would be no other way). And having to remove that file from all the nodes doesn't seem to be that user-friendly either. 

I'm not sure how other possible "unsafe" parameter changes have been handled in the past, but my concern is just avoiding that the prevention turns out to be more problematic than the original issue itself (I mean, changing that parameter will break the connectivity in the whole cluster, but that could be even fine while performing tests).

giorio94 commented on pkg/identity/numericidentity.go:
```diff
@@ -83,20 +84,24 @@ const (
+	// clusterIDShift is the number of bits to shift a cluster ID in a numeric
 	// identity and is equal to the number of bits that represent a cluster-local identity.
-	ClusterIDShift int = NumericIdentityBitlength - ClusterIDLen
+	clusterIDShift int = NumericIdentityBitlength - clusterIDLen
```
Nit. Could be dropped, and computed based on clusterIDLen.

thorn3r replied:
i thought the same as well, but if this is removed we also need to replace `MaximumAllocationIdentity` with a Getter, which will likely then depend on the global `ClusterIDMax`. This area is a bit tangled with the global, I tried a few different approaches before I settled on this one

giorio94 commented on pkg/identity/numericidentity_test.go:
```diff
@@ -79,3 +82,47 @@ func TestAsUint32Slice(t *testing.T) {
+					// skip the current test case itself
+					continue
+				}
+				cinfo.InitClusterIDMax()
```
This is a no-op, because cinfo is not updated.

thorn3r replied:
corrected this one

giorio94 commented on pkg/identity/numericidentity.go:
I'd merge this commit with the one that introduced these variables, for improved clarity. Beides that, I fear that there's still the possibility for possible issues due to ordering. In particular, there's nothing which prevents accessing any of the fields before initializing the ClusterIDMax value. I don't know if there's anything that we can do to improve this though.

thorn3r replied:
I think ideally, these variables and ClusterIDMax should be provided in their own Identity-related hive cell. Took a swing at this, but the refactor was pretty large.

giorio94 replied:
Yeah, that looks out of scope. Could setting the initial values of these variables to non-sensible defaults (like 0) help catching possible misuses? Just to avoid missing that something is later introduced in the wrong place because the default also matches the expected value in all scenarios except that that if extended clustermesh is configured.

thorn3r replied:
i turned the globals into Getters to avoid this issue. The only one left is ClusterIDMax, tbd in another refactor later :sweat_smile: . Added a new commit but can squash it into the other if it looks good, lmk what you think!

giorio94 replied:
Makes sense to me, thanks!

giorio94 reviewed (changes requested):
Thanks @thorn3r for all the changes. I've left a few additional comments inline, but the PR looks almost ready from my POW.

christarazi commented on pkg/datapath/linux/config/config.go:
```diff
@@ -140,6 +132,48 @@ func (h *HeaderfileWriter) WriteClusterConfig(w io.Writer) error {
+		if newVal, ok := cDefinesMap[key]; ok && newVal != currentVal {
+			return fmt.Errorf("cluster_config.h cannot be changed. %s was defined as %s, but new agent configuration would change it to %v", key, currentVal, newVal)
+		}
+	}
```
IIUC, this is validating the cluster config file that comes from clustermesh, each time a new cluster is added to the mesh, right? If that's the case, adding that context somewhere is helpful in tying the high-level of clustermesh down to these details of the code to make it easier for developers to understand the full picture of the code.

thorn3r replied:
This part is actually checking an existing `cluster_config.h` header that's on disk, and making sure its values haven't changed. This is to prevent a user from breaking their cluster if they try to enable `max-connected-clusters=511` on an existing cluster thats running on the default 255. I'll update the comments here to make it more clear

christarazi commented on test/k8s/updates.go:
```diff
@@ -17,6 +17,11 @@ import (
+	clusterName = "test-upgrade-downgrade"
+	clusterID   = "170"
+)
+
```
Why these values? Are they necessary for testing this feature in the ginkgo suite?

thorn3r replied:
Since keys/values for BPF maps were updated to use a uint16 for ClusterID, setting the ClusterID in these tests makes sure that we can upgrade (maps go from uint8 -> uint16) and downgrade (uint16 -> back to uint8) safely without missing any tail calls and dropping connections.
We can't add coverage here for using 'max-connected-clusters' yet because this feature is only available to new clusters (no migration for a running cluster using MCC=255 to MCC=511), but it can be added once the feature is released.
The clusterID value was initially 1, but 170 (10101010) was suggested to cover more of the bitspace

christarazi replied:
A code comment describing that would be great :slightly_smiling_face:

thorn3r replied:
added!

christarazi reviewed (commented):
LGTM from the ci-structure perspective. Just a few clarifying comments. I did not do a full review of the code; just for my codeowners and other areas like the new ConfigReader that popped out at me.

thorn3r committed:
refactor: replace identity allocation globals

MinimalAllocationIdentity and MaximumAllocationIdentity are global
variables that can be changed at runtime by the
InitMinMaxIdentityAllocation function. Since these are dependent on
ClusterIDShift, which is no longer a constant, this commit replaces
the globals with Getter methods to compute the correct value and avoid
a race condition in which a global is accessed before it is correctly
initialized.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

ti-mo commented on pkg/datapath/linux/config/config.go:
```diff
@@ -98,6 +98,83 @@ func writeIncludes(w io.Writer) (int, error) {
+}
+
+// ReadConfigMacros reads the given headerfile and returns a map of the macro definitions.
+func (h *HeaderfileReader) ReadConfigMacros(r io.Reader) map[string]string {
```
As mentioned in https://github.com/cilium/cilium/pull/27520/files#r1346120795, I think this is problematic. We shouldn't treat headers as a state file, nor should we validate user configuration this deep into the code.

thorn3r replied:
thanks for taking another look, Timo

> Wouldn't it be simpler to remove the writable bit from cluster_config.h if we want to prevent future modifications?

Definitely simpler, but I thought this approach would be more flexible, allowing additional macros to be defined or repairing a header with a broken config

> Do we have validating webhooks for configmap updates?

Not currently, but there's an open issue for that [here](https://github.com/cilium/cilium/issues/18652)

> Could you outline the exact issues you ran into adding these values to node_config.h?

I don't recall the exact issues offhand but I think it was related to defining macros in common.h that reference values from node_config.h. node_config.h is only sometimes included indirectly in common.h. I'm sure I could revisit and make that work if using the cluster_config header for validation is out.

It seems like we may not have a precedent for validating a running config against a previous one, and without that we may want to consider shelving this until we implement the migration path for existing clusters.
cc: @aanm @marseel for thoughts on this

ti-mo replied:
Instead of reading the header back out, is there an other oracle you could use to determine if a cluster was deployed with a smaller cluster limit? Cilium already leaves quite a bit of state on the node in the form of maps and files in `/var/run/cilium/state`. There's also `agent-runtime-config.json` in there, for example.

thorn3r replied:
`agent-runtime-config.json` might be a good alternative! seems like I should spend some time exploring what secrets that directory contains :grin:

thorn3r replied:
@ti-mo dropped the validation and cluster_config.h. The new macros are now defined in node_config.h

giorio94 commented on pkg/identity/numericidentity.go:
```diff
@@ -383,22 +383,49 @@ func InitWellKnownIdentities(c Configuration, cinfo cmtypes.ClusterInfo) int {
-	if cinfo.ID > 0 {
+// GetClusterIDLen returns the number of bits that represent a cluster ID in a numeric identity.
+// A sync.Once is used to ensure we only initialize clusterIDShift once.
+
```
Nit. Extra new line

giorio94 commented on pkg/maps/ctmap/per_cluster_ctmap.go:
```diff
@@ -124,17 +122,17 @@ func (v *PerClusterCTMapVal) String() string    { return fmt.Sprintf("fd=%d", v.
 
 // NewPerClusterCTMaps returns a new instance of the per-cluster CT maps manager.
-func NewPerClusterCTMaps(ipv4, ipv6 bool) *perClusterCTMaps {
+func NewPerClusterCTMaps(ipv4, ipv6 bool, outerMapEntries int) *perClusterCTMaps {
```
Sorry, I know that it was me who suggested to make the number of entries a parameter, but upon a more detailed look it is probably better to hardcode it. The reason is that `cmtypes.ClusterIDMax` is already accessed directly once in `cleanup`, and indirectly by `ValidateClusterID`, so that would increase the risk of inconsistencies. The other alternative would be to also refactor the other usages, but that looks out of scope here.

giorio94 commented on pkg/clustermesh/remote_cluster_test.go:
```diff
@@ -234,3 +237,132 @@ func TestIPCacheWatcherOpts(t *testing.T) {
+			assertion: assert.Error,
+		},
+		{
+			name:      "Inalid config (ClusterMesh511)",
```
```suggestion
			name:      "Invalid config (ClusterMesh511)",
```

Nit.

giorio94 commented on pkg/clustermesh/remote_cluster.go:
```diff
@@ -175,7 +175,7 @@ func (rc *remoteCluster) Status() *models.RemoteCluster {
 
 func (rc *remoteCluster) ClusterConfigRequired() bool {
-	return rc.mesh.conf.ConfigValidationMode == types.Strict
+	return (rc.mesh.conf.ConfigValidationMode == types.Strict) || rc.mesh.conf.ClusterInfo.ExtendedClusterMeshEnabled()
```
The `ClusterInfo.ExtendedClusterMeshEnabled` part could be directly moved into `ValidateRemoteConfig`, as it only depends on `ClusterInfo`.

thorn3r replied:
This is true, but I think it might make more sense here if we expect someone may reuse ClusterConfigRequired again in the future. If extended clustermesh is enabled, the config will always be required

thorn3r replied:
ending up going with your suggestion, as after moving the test it introduced a cyclical dependency :grin:

giorio94 commented on pkg/clustermesh/remote_cluster_test.go:
```diff
@@ -234,3 +237,132 @@ func TestIPCacheWatcherOpts(t *testing.T) {
 	}
 }
+
+func TestRemoteClusterValidate(t *testing.T) {
```
This test could be likely moved alongside the `ClusterInfo.ValidateRemoteConfig` definition.

giorio94 reviewed (commented):
Overall looks mostly good to me, only a few nits inline.

The only part I'm not really convinced about is the validation of the header file, which seems pretty prone to plague us in the future. I'd be personally more in favor of leaving that out, and just accept the risk of possibly breaking existing connections if the setting is changed carelessly. That said, I'll defer that to Timo and the others with more experience on that aspect.

giorio94 reviewed (approved):
Approving, pending the discussion about the validation.

giorio94 reviewed (approved):
It seems I've still a pending review even though I approved the PR :confused: Let's try again

thorn3r committed:
bpf tests: reorder dependencies

Since overloadable_skb.h now requires macros defined in node_config.h,
it is necessary to reorder dependencies in some tests that are using
unspec.h and including bpf_sock.c, so that when overloadable_skb.h is
included in lib/common.h, it occurs after bpf_sock.c includes
node_config.h

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

giorio94 commented on pkg/option/config.go:
```diff
@@ -2850,6 +2850,10 @@ func (c *DaemonConfig) validateIPv6NAT46x64CIDR() error {
 func (c *DaemonConfig) Validate(vp *viper.Viper) error {
+	if err := c.validatePreviousConfig(c.StateDir); err != nil {
+		return err
+	}
```
AFAICT, this validation is affected by two possible drawbacks:
* It introduces a backward-compatibility requirement on the `DaemonConfig` struct. Let say for instance that at a certain point we had a `Foo` field of type string, which is then switched to be bool. The unmarshalling will then always fail because the two types are incompatible. I'm not sure how likely such a change could be, or whether others could trigger the same issue, but still worth mentioning.
* We currently store only the `DaemonConfig` configurations, which means that logic will no longer work after the refactoring has been completed, and the parameter has been removed from that struct.

That said, I'm still personally not fully convinced that the complexity introduced by this (or equivalent) validations is actually worth, considering that we don't do anything similar for a range of other functionalities (e.g., switching from tunnel to direct routing) which can equivalently disrupt existing connections.

thorn3r replied:
Yes, good point. We _could_ create a simple type just for validation that only contains fields we're interested in (only MaxConnectedClusters right now). But I'm on board with dropping the validation since this is only a temporary measure (we plan to check allocated identities in the future, but as part of the migration work), it may not be useful anywhere else, and also the fact that we already have other options that have a similar potential to disrupt if changed without discretion.

@aanm are you ok with dropping the validation on this?

thorn3r replied:
dropped :+1:

jrajahalme commented on pkg/identity/numericidentity.go:
```diff
@@ -572,7 +602,7 @@ func (id NumericIdentity) IsReservedIdentity() bool {
 // ClusterID returns the cluster ID associated with the identity
 func (id NumericIdentity) ClusterID() uint32 {
-	return (uint32(id) >> 16) & 0xFF
+	return (uint32(id) >> uint32(GetClusterIDShift())) & cmtypes.ClusterIDMax
```
I see no validation that `cmtypes.ClusterIDMax` be a continuous set of bits like 0xff or 0x1ff.

thorn3r replied:
The validation is handled in [InitClusterIDMax here](https://github.com/cilium/cilium/pull/27520/files#diff-b1c7679f51256f1144fd1acc75c61cff5caba43ec2ef73edf8f02c2e0e0449bcR21-R30) and called in [DaemonConfig.Validate](https://github.com/cilium/cilium/pull/27520/files#diff-8a383aefc352ba8854dbab293a563fecf7f8916983d013ab36f677293996ca3cR2912-R2913). The only acceptable values should be 255 and 511. Lmk if you see any issue there

jrajahalme reviewed (approved):
Small nit on option validation.

asauber reviewed (approved):
Config-related changed LGTM. The requirement that the mark-bits are non-contiguous is interesting. Nice work.

ti-mo commented on test/bpf/elf-demo.c:
```diff
@@ -5,6 +5,7 @@
 #include <bpf/api.h>
 
 #include "lib/utils.h"
+#include <node_config.h>
```
Nit: this seems a little arbitrarily placed. :sweat_smile: Does it need to be in this particular spot?

thorn3r replied:
It needs to be placed before common.h, since common includes overloadable -> overloadable_skb which references macros defined in node_config.h. Somewhat related to https://github.com/cilium/cilium/pull/27520/commits/7add8a773922823193bb2af4fa65aa7b9434a6d8. I can add a comment here explaining this

ti-mo replied:
Not sure if there's any reason against it, but could we put a `#pragma once` in node_config.h and make headers depend directly on the stuff they need instead? Having to worry about include order isn't productive.

thorn3r replied:
I'm not sure, but I can try it

thorn3r replied:
Seems like this is causing some issues in tests, since there are conditional defines in node_config.h. I think there's going to need to be some order of includes one way or another, unless we refactor some parts of node_config.h. I'll try to find some time to look over it more if this will be a blocker

ti-mo replied:
Fair enough, it's a super minor thing.

ti-mo commented on pkg/datapath/linux/config/config.go:
```diff
@@ -791,6 +791,14 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 
+	// Write Identity and ClusterID related macros.
+	cDefinesMap["CLUSTER_ID_MAX"] = fmt.Sprintf("%d", option.Config.MaxConnectedClusters)
+	cDefinesMap["CLUSTER_ID_LEN"] = fmt.Sprintf("%d", identity.GetClusterIDLen())
```
This seems to be unused in the C code.

thorn3r replied:
these are used in [overloadable_skb.h](https://github.com/cilium/cilium/pull/27520/files#diff-64fb9c937b4b4671eb58de176f97781c6e467adf3150f405b9f814e05540260a)

thorn3r replied:
removed

ti-mo commented on pkg/datapath/linux/config/config.go:
```diff
@@ -791,6 +791,14 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
+
+	identityMax := (1 << identity.GetClusterIDShift()) - 1
+	cDefinesMap["IDENTITY_MAX"] = fmt.Sprintf("%d", identityMax)
+	cDefinesMap["IDENTITY_LEN"] = fmt.Sprintf("%d", identity.GetClusterIDShift())
```
Would there be a way of pushing the calculation of IDENTITY_LEN into the datapath? Given the limited amount of valid values for MaxConnectedClusters, you could even hardcode the length based on an input of 255 and 511. Encoding these 255/511 invariants helps with simplicity and readability, as the reader immediately knows which output values to expect.

(also, fewer defines = fewer surprises when we turn these into runtime configurables later :slightly_smiling_face:)

thorn3r replied:
> pushing the calculation of IDENTITY_LEN into the datapath

I sort of thought this was what I was doing :sweat_smile: The agent does the calculation and the value is passed into the datapath with this define. Do you have something else in mind?

As for hardcoding, I was really trying to avoid that. This was initially designed to allow extension beyond 511, but it was suggested we should artificially limit it to 255/511 at least initially.

thorn3r replied:
Taking a closer look at this, we can move all of these except CLUSTER_ID_MAX into helpers in the C code, however we'll still need to keep `GetClusterIDShift()` in numericidentity.go since that is used to get the max/min allocation identity in the agent.
Doing the same calculation in 2 different places to get the same value feels prone to breaking in some unexpected ways. What do you think @ti-mo ?

ti-mo replied:
Makes sense, we can commit to treating it as a config value, it makes sense to ensure the agent and datapath agree on the value. The reason I'm pushing back on this so much is that we want to start converting `#define`s in node_config.h and friends to runtime-provided values, meaning `IDENTITY_LEN` cannot appear in any preprocessing directives. If we merge this as-is, we'll need to convert all preprocessing directives later, with the risk of regressions.
So far most of the values in node_config.h are relatively straightforward; addresses, masks and the like, but the way `IDENTITY_LEN` is used in the C feels a little more intricate.

(By the way, play it safe and use unsigned, fixed-length ints for bitwise ops only, see https://github.com/cilium/cilium/pull/28417. Both in Go and C code, e.g. `GetClusterIDShift` should return a u64 or u32)

To make things easier to convert, I'd suggest this in `node_config.h` instead:

```
DECLARE_CONFIG(u64, CLUSTER_ID_SHIFT, "<some description>")
ASSIGN_CONFIG(CLUSTER_ID_SHIFT, <GetClusterIDShift() value>)
```

(note, I haven't created a helper yet for rendering DECLARE_CONFIG and ASSIGN_CONFIG into node_config.h from Go, please add these next to `defineUint32()` and friends)

Then, to reference it in C code in inlined helpers like `get_identity()`:
`foo = CONFIG(CLUSTER_ID_SHIFT)`

You've already designed your helpers in a way that make this change relatively straightforward. :+1: You'll also need to convert e.g. CLUSTER_ID_UPPER_MASK and MARK_MAGIC_CLUSTER_ID_MASK into inlined helpers since these currently depend on IDENTITY_LEN, which is dynamic.

This will allow someone to come in later and simply remove the `ASSIGN_CONFIG` statement to replace it with a runtime-provided value. Using DECLARE_CONFIG makes CLUSTER_ID_SHIFT show up in BTF info in a specific section as well. For context, the plan is to use code generation to parse this section and emit a single config struct for all config values in the ELF, so we can provide these values in a type-safe manner and effectively bring a datapath config plane to Go.

thorn3r replied:
Done! I used macros defined in clustermesh.h for the helpers so that I could override things that depend on a config load in the dummy node_config.h and bpf tests. Not sure if there's a better way to get the global config macros to play nice, let me know how that looks

ti-mo reviewed (changes requested):
Impressive. Left a few nits, sorry for holding up the process.

thorn3r committed:
replace clustermesh defines with global config var

Remove IDENTITY_MAX from node_config.h and replace IDENTITY_LEN with
a global config variable using DECLARE_CONFIG and ASSIGN_CONFIG. This
will allow replacement at runtime in later iterations.

New helper functions now calculate the other previously defined macros
for IDENTITY_MAX, CLUSTER_ID_UPPER_MASK, and MARK_MAGIC_CLUSTER_ID_MASK,
since these all depend on IDENTITY_LEN, which could change after
compilation.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

ti-mo commented on pkg/clustermesh/types/types.go:
```diff
@@ -4,18 +4,33 @@
-	// ClusterIDMax is the maximum value of the cluster ID
-	ClusterIDMax = 255
+	ClusterIDMin    = 0
+	ClusterIDExt511 = 511
```
Nit: what does `ClusterIDExt511` represent, exactly?

thorn3r replied:
A valid configuration for `--max-connected-clusters`. The flag should accept 255 (defaults.MaxConnectedClusters) or 511

ti-mo commented on pkg/datapath/linux/config/utils.go:
```diff
@@ -60,3 +60,37 @@ func defineMAC(name string, addr []byte) string {
+
+// declareConfig writes the C macro for declaring a global configuration variable that can be
+// modified at runtime.
+func declareConfig(name string, value any, description string) string {
```
Thanks for adding these :+1:

ti-mo commented on bpf/lib/clustermesh.h:
```diff
@@ -1,11 +1,48 @@
+static __always_inline __u32
+get_identity_len()
+{
+	__u32 identity_len = CONFIG(IDENTITY_LEN);
```
I'm slightly surprised this works, since `IDENTITY_LEN` is defined to `get_identity_len()` a few lines above. I think that's due to the CONFIG() macro pasting the argument verbatim during the initial preprocessor pass. I can see this confusing some editors/language servers though.

Since this is a datapath configurable, what about calling this `CONFIG(cluster_id_length)`? Uppercase names imply preprocessor directives, which IDENTITY_LEN technically isn't. When generating Go code, we can nicely transform `cluster_id_length` into `ClusterIDLength`. We're setting the convention for all future configurables here, so worth thinking this through.

thorn3r replied:
yep good point, that makes sense. i'll push that change

thorn3r replied:
updated :+1:

ti-mo replied:
Thanks, looks good!

ti-mo reviewed (approved):
Thanks for bearing with me! Left a few nits, good to go otherwise. :rocket:
